File: cc-native-compiler-osaka-202605/langs/typescript/src/codegen.ts
import type {
  Expr, FunDef, Param, Program, Stmt, Type,
} from "./ast.ts";

type FunSig    = { params: Param[]; retType: Type };
type VarSlot   = { ptr: string; ty: Type };
type StmtRes   = { terminated: boolean };
type ExprRes   = { value: string; ty: Type };

class CodeGen {
  private globalDecls: string[] = [];
  private funcDefs:    string[] = [];
  private mainBuf:     string[] = [];

  private regCounter    = 0;
  private labelCounter  = 0;
  private strCounter    = 0;
  private suffixCounter = 0;

  private funcs = new Map<string, FunSig>();
  private strPool = new Map<string, string>();

  generate(program: Program): string {
    for (const f of program.funcs) {
      this.funcs.set(f.name, { params: f.params, retType: f.retType });
    }
    for (const f of program.funcs) this.genFunction(f);
    this.genMain(program.topStmts);
    return this.result();
  }

  // === helpers ===

  private freshReg(prefix: string): string {
    this.regCounter++;
    return `%${prefix}.${this.regCounter}`;
  }
  private freshLabel(prefix: string): string {
    this.labelCounter++;
    return `${prefix}.${this.labelCounter}`;
  }
  private freshSuffix(): string {
    this.suffixCounter++;
    return String(this.suffixCounter);
  }

  private llvmType(t: Type): string {
    switch (t.kind) {
      case "TInt":    return "i64";
      case "TString": return "ptr";
      case "TList":   return "ptr";
    }
  }

  private defaultRet(t: Type): string {
    switch (t.kind) {
      case "TInt":    return "ret i64 0";
      case "TString": return "ret ptr null";
      case "TList":   return "ret ptr null";
    }
  }

  // === 文字列リテラル ===

  private addString(s: string): string {
    const existing = this.strPool.get(s);
    if (existing !== undefined) return existing;
    this.strCounter++;
    const name = `@.str.${this.strCounter}`;
    const enc = encodeBytes(s);
    const totalLen = enc.byteLen + 1;
    this.globalDecls.push(
      `${name} = private constant [${totalLen} x i8] c"${enc.escaped}\\00"`,
    );
    this.strPool.set(s, name);
    return name;
  }

  // === エントリ ===

  private genMain(stmts: Stmt[]): void {
    this.regCounter = 0;
    this.labelCounter = 0;
    this.mainBuf.push("define i32 @main() {");
    this.mainBuf.push("entry:");
    const env = new Map<string, VarSlot>();
    const r = this.genStmts(stmts, env, this.mainBuf);
    if (!r.terminated) {
      this.mainBuf.push("  ret i32 0");
    }
    this.mainBuf.push("}");
    this.mainBuf.push("");
  }

  private genFunction(f: FunDef): void {
    this.regCounter = 0;
    this.labelCounter = 0;
    const paramStr = f.params
      .map((p) => `${this.llvmType(p.ty)} %arg.${p.name}`)
      .join(", ");
    this.funcDefs.push(`define ${this.llvmType(f.retType)} @${f.name}(${paramStr}) {`);
    this.funcDefs.push("entry:");
    const env = new Map<string, VarSlot>();
    for (const p of f.params) {
      const ptr = `%${p.name}.addr`;
      this.funcDefs.push(`  ${ptr} = alloca ${this.llvmType(p.ty)}`);
      this.funcDefs.push(`  store ${this.llvmType(p.ty)} %arg.${p.name}, ptr ${ptr}`);
      env.set(p.name, { ptr, ty: p.ty });
    }
    const r = this.genStmts(f.body, env, this.funcDefs);
    if (!r.terminated) {
      this.funcDefs.push(`  ${this.defaultRet(f.retType)}`);
    }
    this.funcDefs.push("}");
    this.funcDefs.push("");
  }

  // === 文の生成 ===

  private genStmts(stmts: Stmt[], env: Map<string, VarSlot>, buf: string[]): StmtRes {
    let terminated = false;
    for (const s of stmts) {
      if (terminated) break;
      terminated = this.genStmt(s, env, buf);
    }
    return { terminated };
  }

  private genStmt(s: Stmt, env: Map<string, VarSlot>, buf: string[]): boolean {
    switch (s.kind) {
      case "ValDef":
      case "VarDef": {
        const r = this.genExpr(s.init, env, buf);
        const ptr = `%${s.name}.addr.${this.freshSuffix()}`;
        buf.push(`  ${ptr} = alloca ${this.llvmType(r.ty)}`);
        buf.push(`  store ${this.llvmType(r.ty)} ${r.value}, ptr ${ptr}`);
        env.set(s.name, { ptr, ty: r.ty });
        return false;
      }
      case "Reassign": {
        const r = this.genExpr(s.value, env, buf);
        const slot = env.get(s.name)!;
        buf.push(`  store ${this.llvmType(slot.ty)} ${r.value}, ptr ${slot.ptr}`);
        return false;
      }
      case "If": {
        const c = this.genExpr(s.cond, env, buf);
        const condBool = this.freshReg("c");
        buf.push(`  ${condBool} = icmp ne i64 ${c.value}, 0`);
        const thenLbl = this.freshLabel("then");
        const elseLbl = this.freshLabel("else");
        const endLbl  = this.freshLabel("ifend");
        buf.push(`  br i1 ${condBool}, label %${thenLbl}, label %${elseLbl}`);

        buf.push(`${thenLbl}:`);
        const thenR = this.genStmts(s.thenBlk, new Map(env), buf);
        if (!thenR.terminated) buf.push(`  br label %${endLbl}`);

        buf.push(`${elseLbl}:`);
        const elseR = this.genStmts(s.elseBlk, new Map(env), buf);
        if (!elseR.terminated) buf.push(`  br label %${endLbl}`);

        if (thenR.terminated && elseR.terminated) {
          buf.push(`${endLbl}:`);
          buf.push("  unreachable");
          return true;
        }
        buf.push(`${endLbl}:`);
        return false;
      }
      case "While": {
        const condLbl = this.freshLabel("wcond");
        const bodyLbl = this.freshLabel("wbody");
        const endLbl  = this.freshLabel("wend");
        buf.push(`  br label %${condLbl}`);
        buf.push(`${condLbl}:`);
        const c = this.genExpr(s.cond, env, buf);
        const condBool = this.freshReg("c");
        buf.push(`  ${condBool} = icmp ne i64 ${c.value}, 0`);
        buf.push(`  br i1 ${condBool}, label %${bodyLbl}, label %${endLbl}`);
        buf.push(`${bodyLbl}:`);
        const bodyR = this.genStmts(s.body, new Map(env), buf);
        if (!bodyR.terminated) buf.push(`  br label %${condLbl}`);
        buf.push(`${endLbl}:`);
        return false;
      }
      case "Print": {
        const r = this.genExpr(s.value, env, buf);
        switch (r.ty.kind) {
          case "TInt":
            buf.push(`  call i32 (ptr, ...) @printf(ptr @.fmt.int, i64 ${r.value})`);
            break;
          case "TString":
            buf.push(`  call i32 (ptr, ...) @printf(ptr @.fmt.str, ptr ${r.value})`);
            break;
          case "TList": {
            const ph = this.addString("[list]");
            buf.push(`  call i32 (ptr, ...) @printf(ptr @.fmt.str, ptr ${ph})`);
            break;
          }
        }
        return false;
      }
      case "Return": {
        const r = this.genExpr(s.value, env, buf);
        buf.push(`  ret ${this.llvmType(r.ty)} ${r.value}`);
        return true;
      }
      case "ExprStmt":
        this.genExpr(s.value, env, buf);
        return false;
    }
  }

  // === 式の生成 ===

  private genExpr(e: Expr, env: Map<string, VarSlot>, buf: string[]): ExprRes {
    switch (e.kind) {
      case "IntLit":
        return { value: e.value.toString(), ty: { kind: "TInt" } };
      case "StrLit":
        return { value: this.addString(e.value), ty: { kind: "TString" } };
      case "Var": {
        const slot = env.get(e.name)!;
        const r = this.freshReg("v");
        buf.push(`  ${r} = load ${this.llvmType(slot.ty)}, ptr ${slot.ptr}`);
        return { value: r, ty: slot.ty };
      }
      case "Neg": {
        const inner = this.genExpr(e.inner, env, buf);
        const r = this.freshReg("neg");
        buf.push(`  ${r} = sub i64 0, ${inner.value}`);
        return { value: r, ty: { kind: "TInt" } };
      }
      case "BinOp": {
        const l = this.genExpr(e.left, env, buf);
        const r = this.genExpr(e.right, env, buf);
        switch (e.op) {
          case "+":  return this.emitArith("add",  l.value, r.value, buf);
          case "-":  return this.emitArith("sub",  l.value, r.value, buf);
          case "*":  return this.emitArith("mul",  l.value, r.value, buf);
          case "/":  return this.emitArith("sdiv", l.value, r.value, buf);
          case "%":  return this.emitArith("srem", l.value, r.value, buf);
          case "==": return this.emitCmp("eq",  l.value, r.value, buf);
          case "!=": return this.emitCmp("ne",  l.value, r.value, buf);
          case "<":  return this.emitCmp("slt", l.value, r.value, buf);
          case "<=": return this.emitCmp("sle", l.value, r.value, buf);
          case ">":  return this.emitCmp("sgt", l.value, r.value, buf);
          case ">=": return this.emitCmp("sge", l.value, r.value, buf);
          default:   throw new Error(`bad op ${e.op}`);
        }
      }
      case "Call": {
        const sig = this.funcs.get(e.name);
        if (!sig) throw new Error(`unknown function: ${e.name}`);
        const argParts: string[] = [];
        for (let i = 0; i < e.args.length; i++) {
          const av = this.genExpr(e.args[i]!, env, buf);
          const pty = sig.params[i]!.ty;
          argParts.push(`${this.llvmType(pty)} ${av.value}`);
        }
        const r = this.freshReg("call");
        buf.push(`  ${r} = call ${this.llvmType(sig.retType)} @${e.name}(${argParts.join(", ")})`);
        return { value: r, ty: sig.retType };
      }
      case "Index": {
        const arr = this.genExpr(e.arr, env, buf);
        const idx = this.genExpr(e.idx, env, buf);
        if (arr.ty.kind !== "TList") throw new Error(`cannot index non-list`);
        const elemTy = arr.ty.elem;
        const ep = this.freshReg("ep");
        buf.push(`  ${ep} = getelementptr ${this.llvmType(elemTy)}, ptr ${arr.value}, i64 ${idx.value}`);
        const r = this.freshReg("elem");
        buf.push(`  ${r} = load ${this.llvmType(elemTy)}, ptr ${ep}`);
        return { value: r, ty: elemTy };
      }
      case "ListLit": {
        if (e.elems.length === 0) throw new Error("empty list not supported");
        const first = this.genExpr(e.elems[0]!, env, buf);
        const elemTy = first.ty;
        const byteSize = e.elems.length * 8;
        const basePtr = this.freshReg("list");
        buf.push(`  ${basePtr} = call ptr @malloc(i64 ${byteSize})`);
        const ep0 = this.freshReg("ep");
        buf.push(`  ${ep0} = getelementptr ${this.llvmType(elemTy)}, ptr ${basePtr}, i64 0`);
        buf.push(`  store ${this.llvmType(elemTy)} ${first.value}, ptr ${ep0}`);
        for (let i = 1; i < e.elems.length; i++) {
          const v = this.genExpr(e.elems[i]!, env, buf);
          const ep = this.freshReg("ep");
          buf.push(`  ${ep} = getelementptr ${this.llvmType(elemTy)}, ptr ${basePtr}, i64 ${i}`);
          buf.push(`  store ${this.llvmType(elemTy)} ${v.value}, ptr ${ep}`);
        }
        return { value: basePtr, ty: { kind: "TList", elem: elemTy } };
      }
    }
  }

  private emitArith(op: string, lv: string, rv: string, buf: string[]): ExprRes {
    const out = this.freshReg("b");
    buf.push(`  ${out} = ${op} i64 ${lv}, ${rv}`);
    return { value: out, ty: { kind: "TInt" } };
  }

  private emitCmp(pred: string, lv: string, rv: string, buf: string[]): ExprRes {
    const bool = this.freshReg("cmp");
    buf.push(`  ${bool} = icmp ${pred} i64 ${lv}, ${rv}`);
    const ext = this.freshReg("ext");
    buf.push(`  ${ext} = zext i1 ${bool} to i64`);
    return { value: ext, ty: { kind: "TInt" } };
  }

  // === 最終結合 ===

  private result(): string {
    const out: string[] = [];
    out.push("; ModuleID = 'nb-lang'");
    out.push(`target triple = "${targetTriple()}"`);
    out.push("");
    out.push(`@.fmt.int = private constant [5 x i8] c"%ld\\0A\\00"`);
    out.push(`@.fmt.str = private constant [4 x i8] c"%s\\0A\\00"`);
    out.push("declare i32 @printf(ptr, ...)");
    out.push("declare ptr @malloc(i64)");
    out.push("");
    out.push(...this.globalDecls);
    out.push("");
    out.push(...this.funcDefs);
    out.push(...this.mainBuf);
    return out.join("\n");
  }
}

function targetTriple(): string {
  return process.platform === "darwin"
    ? "arm64-apple-macosx15.0.0"
    : "x86_64-pc-linux-gnu";
}

function encodeBytes(s: string): { escaped: string; byteLen: number } {
  const utf8 = new TextEncoder().encode(s);
  let escaped = "";
  for (const b of utf8) {
    const c = b & 0xff;
    if (c >= 0x20 && c <= 0x7e && c !== 0x22 /* " */ && c !== 0x5c /* \ */) {
      escaped += String.fromCharCode(c);
    } else {
      escaped += "\\" + c.toString(16).toUpperCase().padStart(2, "0");
    }
  }
  return { escaped, byteLen: utf8.length };
}

export function generate(program: Program): string {
  return new CodeGen().generate(program);
}

File: cc-native-compiler-osaka-202605/langs/java/src/main/java/nblang/CodeGen.java
package nblang;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import nblang.Ast.Expr;
import nblang.Ast.FunDef;
import nblang.Ast.Param;
import nblang.Ast.Program;
import nblang.Ast.Stmt;
import nblang.Ast.Type;

public final class CodeGen {
    private record FunSig(List<Param> params, Type retType) {}
    private record VarSlot(String ptr, Type ty) {}
    private record EncodedBytes(String escaped, int byteLen) {}
    private record StmtResult(Map<String, VarSlot> env, boolean terminated) {}
    private record ExprResult(String value, Type ty) {}

    private final StringBuilder globalDecls = new StringBuilder();
    private final StringBuilder funcDefs    = new StringBuilder();
    private final StringBuilder mainBuf     = new StringBuilder();

    private int regCounter    = 0;
    private int labelCounter  = 0;
    private int strCounter    = 0;
    private int suffixCounter = 0;

    private Map<String, FunSig> funcs = Map.of();
    private final Map<String, String> strPool = new LinkedHashMap<>();

    private CodeGen() {}

    public static String generate(Program program) {
        CodeGen gen = new CodeGen();
        gen.gen(program);
        return gen.result();
    }

    // === helpers ===

    private String freshReg(String prefix) {
        regCounter++;
        return "%" + prefix + "." + regCounter;
    }

    private String freshLabel(String prefix) {
        labelCounter++;
        return prefix + "." + labelCounter;
    }

    private String freshSuffix() {
        suffixCounter++;
        return Integer.toString(suffixCounter);
    }

    private static String llvmType(Type t) {
        return switch (t) {
            case Type.TInt ignored    -> "i64";
            case Type.TString ignored -> "ptr";
            case Type.TList ignored   -> "ptr";
        };
    }

    private static String defaultRet(Type t) {
        return switch (t) {
            case Type.TInt ignored    -> "ret i64 0";
            case Type.TString ignored -> "ret ptr null";
            case Type.TList ignored   -> "ret ptr null";
        };
    }

    // === 文字列リテラル ===

    private String addString(String s) {
        String existing = strPool.get(s);
        if (existing != null) return existing;
        strCounter++;
        String name = "@.str." + strCounter;
        EncodedBytes enc = encodeBytes(s);
        int totalLen = enc.byteLen() + 1;
        globalDecls.append(name)
                .append(" = private constant [")
                .append(totalLen)
                .append(" x i8] c\"")
                .append(enc.escaped())
                .append("\\00\"\n");
        strPool.put(s, name);
        return name;
    }

    private static EncodedBytes encodeBytes(String s) {
        StringBuilder sb = new StringBuilder();
        byte[] utf8 = s.getBytes(StandardCharsets.UTF_8);
        for (byte b : utf8) {
            int c = b & 0xff;
            if (c >= 0x20 && c <= 0x7e && c != '"' && c != '\\') {
                sb.append((char) c);
            } else {
                sb.append(String.format("\\%02X", c));
            }
        }
        return new EncodedBytes(sb.toString(), utf8.length);
    }

    // === エントリ ===

    private void gen(Program program) {
        Map<String, FunSig> sigs = new HashMap<>();
        for (FunDef f : program.funcs()) {
            sigs.put(f.name(), new FunSig(f.params(), f.retType()));
        }
        funcs = sigs;

        for (FunDef f : program.funcs()) {
            genFunction(f);
        }
        genMain(program.topStmts());
    }

    private void genMain(List<Stmt> stmts) {
        regCounter = 0;
        labelCounter = 0;
        mainBuf.append("define i32 @main() {\n");
        mainBuf.append("entry:\n");
        Map<String, VarSlot> env = new HashMap<>();
        StmtResult r = genStmts(stmts, env, mainBuf);
        if (!r.terminated()) {
            mainBuf.append("  ret i32 0\n");
        }
        mainBuf.append("}\n\n");
    }

    private void genFunction(FunDef f) {
        regCounter = 0;
        labelCounter = 0;
        StringBuilder paramStr = new StringBuilder();
        for (int i = 0; i < f.params().size(); i++) {
            Param p = f.params().get(i);
            if (i > 0) paramStr.append(", ");
            paramStr.append(llvmType(p.ty())).append(" %arg.").append(p.name());
        }
        funcDefs.append("define ").append(llvmType(f.retType()))
                .append(" @").append(f.name())
                .append("(").append(paramStr).append(") {\n");
        funcDefs.append("entry:\n");
        Map<String, VarSlot> env = new HashMap<>();
        for (Param p : f.params()) {
            String ptr = "%" + p.name() + ".addr";
            funcDefs.append("  ").append(ptr)
                    .append(" = alloca ").append(llvmType(p.ty())).append("\n");
            funcDefs.append("  store ").append(llvmType(p.ty()))
                    .append(" %arg.").append(p.name())
                    .append(", ptr ").append(ptr).append("\n");
            env.put(p.name(), new VarSlot(ptr, p.ty()));
        }
        StmtResult r = genStmts(f.body(), env, funcDefs);
        if (!r.terminated()) {
            funcDefs.append("  ").append(defaultRet(f.retType())).append("\n");
        }
        funcDefs.append("}\n\n");
    }

    // === 文の生成 ===
    private StmtResult genStmts(List<Stmt> stmts, Map<String, VarSlot> env, StringBuilder buf) {
        boolean terminated = false;
        int idx = 0;
        while (idx < stmts.size() && !terminated) {
            terminated = genStmt(stmts.get(idx), env, buf);
            idx++;
        }
        return new StmtResult(env, terminated);
    }

    private boolean genStmt(Stmt s, Map<String, VarSlot> env, StringBuilder buf) {
        return switch (s) {
            case Stmt.ValDef v -> {
                ExprResult r = genExpr(v.init(), env, buf);
                String ptr = "%" + v.name() + ".addr." + freshSuffix();
                buf.append("  ").append(ptr).append(" = alloca ").append(llvmType(r.ty())).append("\n");
                buf.append("  store ").append(llvmType(r.ty()))
                        .append(" ").append(r.value())
                        .append(", ptr ").append(ptr).append("\n");
                env.put(v.name(), new VarSlot(ptr, r.ty()));
                yield false;
            }
            case Stmt.VarDef v -> {
                ExprResult r = genExpr(v.init(), env, buf);
                String ptr = "%" + v.name() + ".addr." + freshSuffix();
                buf.append("  ").append(ptr).append(" = alloca ").append(llvmType(r.ty())).append("\n");
                buf.append("  store ").append(llvmType(r.ty()))
                        .append(" ").append(r.value())
                        .append(", ptr ").append(ptr).append("\n");
                env.put(v.name(), new VarSlot(ptr, r.ty()));
                yield false;
            }
            case Stmt.Reassign re -> {
                ExprResult r = genExpr(re.value(), env, buf);
                VarSlot slot = env.get(re.name());
                buf.append("  store ").append(llvmType(slot.ty()))
                        .append(" ").append(r.value())
                        .append(", ptr ").append(slot.ptr()).append("\n");
                yield false;
            }
            case Stmt.IfS iff -> {
                ExprResult c = genExpr(iff.cond(), env, buf);
                String condBool = freshReg("c");
                buf.append("  ").append(condBool).append(" = icmp ne i64 ").append(c.value()).append(", 0\n");
                String thenLbl = freshLabel("then");
                String elseLbl = freshLabel("else");
                String endLbl  = freshLabel("ifend");
                buf.append("  br i1 ").append(condBool)
                        .append(", label %").append(thenLbl)
                        .append(", label %").append(elseLbl).append("\n");

                buf.append(thenLbl).append(":\n");
                StmtResult thenR = genStmts(iff.thenBlk(), new HashMap<>(env), buf);
                if (!thenR.terminated()) buf.append("  br label %").append(endLbl).append("\n");

                buf.append(elseLbl).append(":\n");
                StmtResult elseR = genStmts(iff.elseBlk(), new HashMap<>(env), buf);
                if (!elseR.terminated()) buf.append("  br label %").append(endLbl).append("\n");

                if (thenR.terminated() && elseR.terminated()) {
                    buf.append(endLbl).append(":\n");
                    buf.append("  unreachable\n");
                    yield true;
                } else {
                    buf.append(endLbl).append(":\n");
                    yield false;
                }
            }
            case Stmt.WhileS w -> {
                String condLbl = freshLabel("wcond");
                String bodyLbl = freshLabel("wbody");
                String endLbl  = freshLabel("wend");
                buf.append("  br label %").append(condLbl).append("\n");
                buf.append(condLbl).append(":\n");
                ExprResult c = genExpr(w.cond(), env, buf);
                String condBool = freshReg("c");
                buf.append("  ").append(condBool).append(" = icmp ne i64 ").append(c.value()).append(", 0\n");
                buf.append("  br i1 ").append(condBool)
                        .append(", label %").append(bodyLbl)
                        .append(", label %").append(endLbl).append("\n");
                buf.append(bodyLbl).append(":\n");
                StmtResult bodyR = genStmts(w.body(), new HashMap<>(env), buf);
                if (!bodyR.terminated()) buf.append("  br label %").append(condLbl).append("\n");
                buf.append(endLbl).append(":\n");
                yield false;
            }
            case Stmt.Print p -> {
                ExprResult r = genExpr(p.value(), env, buf);
                switch (r.ty()) {
                    case Type.TInt ignored ->
                        buf.append("  call i32 (ptr, ...) @printf(ptr @.fmt.int, i64 ")
                                .append(r.value()).append(")\n");
                    case Type.TString ignored ->
                        buf.append("  call i32 (ptr, ...) @printf(ptr @.fmt.str, ptr ")
                                .append(r.value()).append(")\n");
                    case Type.TList ignored -> {
                        String ph = addString("[list]");
                        buf.append("  call i32 (ptr, ...) @printf(ptr @.fmt.str, ptr ")
                                .append(ph).append(")\n");
                    }
                }
                yield false;
            }
            case Stmt.Return ret -> {
                ExprResult r = genExpr(ret.value(), env, buf);
                buf.append("  ret ").append(llvmType(r.ty()))
                        .append(" ").append(r.value()).append("\n");
                yield true;
            }
            case Stmt.ExprStmt es -> {
                genExpr(es.value(), env, buf);
                yield false;
            }
        };
    }

    // === 式の生成 ===
    private ExprResult genExpr(Expr e, Map<String, VarSlot> env, StringBuilder buf) {
        return switch (e) {
            case Expr.IntLit i -> new ExprResult(Long.toString(i.value()), new Type.TInt());
            case Expr.StrLit s -> new ExprResult(addString(s.value()), new Type.TString());
            case Expr.Var v -> {
                VarSlot slot = env.get(v.name());
                String r = freshReg("v");
                buf.append("  ").append(r).append(" = load ").append(llvmType(slot.ty()))
                        .append(", ptr ").append(slot.ptr()).append("\n");
                yield new ExprResult(r, slot.ty());
            }
            case Expr.Neg neg -> {
                ExprResult inner = genExpr(neg.inner(), env, buf);
                String r = freshReg("neg");
                buf.append("  ").append(r).append(" = sub i64 0, ").append(inner.value()).append("\n");
                yield new ExprResult(r, new Type.TInt());
            }
            case Expr.BinOp b -> {
                ExprResult l = genExpr(b.left(), env, buf);
                ExprResult r = genExpr(b.right(), env, buf);
                yield switch (b.op()) {
                    case "+" -> emitArith("add",  l.value(), r.value(), buf);
                    case "-" -> emitArith("sub",  l.value(), r.value(), buf);
                    case "*" -> emitArith("mul",  l.value(), r.value(), buf);
                    case "/" -> emitArith("sdiv", l.value(), r.value(), buf);
                    case "%" -> emitArith("srem", l.value(), r.value(), buf);
                    case "==" -> emitCmp("eq",  l.value(), r.value(), buf);
                    case "!=" -> emitCmp("ne",  l.value(), r.value(), buf);
                    case "<"  -> emitCmp("slt", l.value(), r.value(), buf);
                    case "<=" -> emitCmp("sle", l.value(), r.value(), buf);
                    case ">"  -> emitCmp("sgt", l.value(), r.value(), buf);
                    case ">=" -> emitCmp("sge", l.value(), r.value(), buf);
                    default -> throw new RuntimeException("bad op " + b.op());
                };
            }
            case Expr.Call c -> {
                FunSig sig = funcs.get(c.name());
                if (sig == null) throw new RuntimeException("unknown function: " + c.name());
                StringBuilder args = new StringBuilder();
                for (int i = 0; i < c.args().size(); i++) {
                    ExprResult av = genExpr(c.args().get(i), env, buf);
                    Type pty = sig.params().get(i).ty();
                    if (i > 0) args.append(", ");
                    args.append(llvmType(pty)).append(" ").append(av.value());
                }
                String r = freshReg("call");
                buf.append("  ").append(r).append(" = call ").append(llvmType(sig.retType()))
                        .append(" @").append(c.name()).append("(").append(args).append(")\n");
                yield new ExprResult(r, sig.retType());
            }
            case Expr.Index idx -> {
                ExprResult arr = genExpr(idx.arr(), env, buf);
                ExprResult ix  = genExpr(idx.idx(), env, buf);
                Type elemTy;
                if (arr.ty() instanceof Type.TList tl) elemTy = tl.elem();
                else throw new RuntimeException("cannot index non-list: " + arr.ty());
                String ep = freshReg("ep");
                buf.append("  ").append(ep).append(" = getelementptr ").append(llvmType(elemTy))
                        .append(", ptr ").append(arr.value())
                        .append(", i64 ").append(ix.value()).append("\n");
                String r = freshReg("elem");
                buf.append("  ").append(r).append(" = load ").append(llvmType(elemTy))
                        .append(", ptr ").append(ep).append("\n");
                yield new ExprResult(r, elemTy);
            }
            case Expr.ListLit ll -> {
                if (ll.elems().isEmpty()) throw new RuntimeException("empty list not supported");
                ExprResult first = genExpr(ll.elems().get(0), env, buf);
                Type elemTy = first.ty();
                int byteSize = ll.elems().size() * 8;
                String basePtr = freshReg("list");
                buf.append("  ").append(basePtr).append(" = call ptr @malloc(i64 ").append(byteSize).append(")\n");
                String ep0 = freshReg("ep");
                buf.append("  ").append(ep0).append(" = getelementptr ").append(llvmType(elemTy))
                        .append(", ptr ").append(basePtr).append(", i64 0\n");
                buf.append("  store ").append(llvmType(elemTy)).append(" ").append(first.value())
                        .append(", ptr ").append(ep0).append("\n");
                for (int i = 1; i < ll.elems().size(); i++) {
                    ExprResult v = genExpr(ll.elems().get(i), env, buf);
                    String ep = freshReg("ep");
                    buf.append("  ").append(ep).append(" = getelementptr ").append(llvmType(elemTy))
                            .append(", ptr ").append(basePtr).append(", i64 ").append(i).append("\n");
                    buf.append("  store ").append(llvmType(elemTy)).append(" ").append(v.value())
                            .append(", ptr ").append(ep).append("\n");
                }
                yield new ExprResult(basePtr, new Type.TList(elemTy));
            }
        };
    }

    private ExprResult emitArith(String op, String lv, String rv, StringBuilder buf) {
        String out = freshReg("b");
        buf.append("  ").append(out).append(" = ").append(op).append(" i64 ")
                .append(lv).append(", ").append(rv).append("\n");
        return new ExprResult(out, new Type.TInt());
    }

    private ExprResult emitCmp(String pred, String lv, String rv, StringBuilder buf) {
        String bool = freshReg("cmp");
        buf.append("  ").append(bool).append(" = icmp ").append(pred).append(" i64 ")
                .append(lv).append(", ").append(rv).append("\n");
        String ext = freshReg("ext");
        buf.append("  ").append(ext).append(" = zext i1 ").append(bool).append(" to i64\n");
        return new ExprResult(ext, new Type.TInt());
    }

    // === 最終結合 ===

    private String result() {
        StringBuilder sb = new StringBuilder();
        sb.append("; ModuleID = 'nb-lang'\n");
        sb.append("target triple = \"").append(targetTriple()).append("\"\n\n");
        sb.append("@.fmt.int = private constant [5 x i8] c\"%ld\\0A\\00\"\n");
        sb.append("@.fmt.str = private constant [4 x i8] c\"%s\\0A\\00\"\n");
        sb.append("declare i32 @printf(ptr, ...)\n");
        sb.append("declare ptr @malloc(i64)\n\n");
        sb.append(globalDecls);
        sb.append("\n");
        sb.append(funcDefs);
        sb.append(mainBuf);
        return sb.toString();
    }

    private static String targetTriple() {
        String os = System.getProperty("os.name", "").toLowerCase();
        if (os.startsWith("mac")) return "arm64-apple-macosx15.0.0";
        return "x86_64-pc-linux-gnu";
    }
}
